staging/src/k8s.io/apimachinery/pkg/util/net/interface.go

```diff
@@ -96,171 +97,176 @@ func isInterfaceUp(intf *net.Interface) bool {
+	return nil, fmt.Errorf("no acceptable interface from host (last cause: %s).", skipReason)
 }
 
-//ChooseHostInterface is a method used fetch an IP for a daemon.
-//It uses data from /proc/net/route file.
-//For a node with no internet connection ,it returns error
-//For a multi n/w interface node it returns the IP of the interface with gateway on it.
+// ChooseHostInterface attempts to determine the IP for the node.
+// It uses routing table data to find the interface with a default GW,
+// checks if it is up, and returns IP from interface that is not a loopback
+// or LL unicast address. If this fails, it will look through all system
+// interfaces. If no interface is found (e.g. not connected to internet),
+// an error is returned.
 func ChooseHostInterface() (net.IP, error) {
-	inFile, err := os.Open("/proc/net/route")
-	if err != nil {
-		if os.IsNotExist(err) {
-			return chooseHostInterfaceNativeGo()
-		}
-		return nil, err
+	nw := networkInterface{}
+	ip, err := chooseHostInterfaceFromRoute(false, nw)
+	if err == nil {
+		return ip, nil
 	}
-	defer inFile.Close()
-	var nw networkInterfacer = networkInterface{}
-	return chooseHostInterfaceFromRoute(inFile, nw)
+	return chooseIPFromHostInterfaces(false, nw)
 }
 
 type networkInterfacer interface {
 	InterfaceByName(intfName string) (*net.Interface, error)
 	Addrs(intf *net.Interface) ([]net.Addr, error)
+	Interfaces() ([]net.Interface, error)
 }
 
 type networkInterface struct{}
 
 func (_ networkInterface) InterfaceByName(intfName string) (*net.Interface, error) {
-	intf, err := net.InterfaceByName(intfName)
-	if err != nil {
-		return nil, err
-	}
-	return intf, nil
+	return net.InterfaceByName(intfName)
 }
 
 func (_ networkInterface) Addrs(intf *net.Interface) ([]net.Addr, error) {
-	addrs, err := intf.Addrs()
+	return intf.Addrs()
+}
+
+func (_ networkInterface) Interfaces() ([]net.Interface, error) {
+	return net.Interfaces()
+}
+
+func chooseHostInterfaceFromRoute(include_ipv6 bool, nw networkInterfacer) (net.IP, error) {
+	cmd := buildRouteCommand(include_ipv6)
+	raw, err := getDefaultRoutes(cmd)
 	if err != nil {
 		return nil, err
 	}
-	return addrs, nil
-}
-
-func chooseHostInterfaceFromRoute(inFile io.Reader, nw networkInterfacer) (net.IP, error) {
-	routes, err := getRoutes(inFile)
+	routes, err := parseDefaultRoutes(raw)
 	if err != nil {
 		return nil, err
 	}
-	zero := net.IP{0, 0, 0, 0}
-	var finalIP net.IP
-	for i := range routes {
-		//find interface with gateway
-		if routes[i].Destination.Equal(zero) {
-			glog.V(4).Infof("Default route transits interface %q", routes[i].Interface)
-			finalIP, err := getIPFromInterface(routes[i].Interface, nw)
-			if err != nil {
-				return nil, err
-			}
-			if finalIP != nil {
-				glog.V(4).Infof("Choosing IP %v ", finalIP)
-				return finalIP, nil
-			}
+
+	for _, route := range routes {
+		glog.V(4).Infof("Default route transits interface %q", route.Interface)
+		ip, err := getIPFromInterface(route.Interface, nw, include_ipv6)
```

Comment:
If backwards compatibility is a concern, and if we want to avoid imposing configuration complexities, then I think we do need some logic to implement precedence. This would be my preference, maybe others in the community can chime in. If we can agree on an order of precedence (e.g. IPv4 w/ def route, IPv4 w/o def route, IPv6 w/ def route, IPv6 w/o def route), then the implementation shouldn't be too complicated.